triedb/pathdb/buffer.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2022 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package pathdb
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/VictoriaMetrics/fastcache"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+)
+
+// buffer is a collection of modified states along with the modified trie nodes.
```

Comment:
Is there any relation between this buffer implementation, and the one PR:ed in https://github.com/ethereum/go-ethereum/pull/30464 ? Or rather, what is the relation between the two? Orthogonal, or one is an evolution of the other?

Reply:
The buffer here contains (a) trie node data (b) state data, it's still a singleton structure for aggregating db writes;

The #30464 introduces an notion called "frozen" buffer. There are two buffers (totally equal) in the pathdb. If the first buffer is full, it will be scheduled for flushing in the background and once it's fully flushed. In the same time, the new db writes will be accumulated in the second buffer to not block the system.